440 SessionTimeout: Client Session Expired Must Log In Again

grepsuzette opened the pull request:
Usage is not widespread as this is a code used by Microsoft.

See https://stackoverflow.com/questions/1653493/what-http-status-code-is-supposed-to-be-used-to-tell-the-client-the-session-has/22484262#22484262
Or https://en.wikipedia.org/wiki/List_of_HTTP_status_codes#440

However, given the frequent usage of Promises with tink,
seems to me a good idea to have this error code, for better *discrimination*.

For instance, if one was to sanitize http error messages when passing in Production (erase message and override the default message from the error code), having a mere 401 would be problematic, while a 440 would make things clear.

Commit:
440 SessionTimeout: Client Session Expired Must Log In Again

Usage is not widespread as this is a code used by Microsoft.
https://stackoverflow.com/questions/1653493/what-http-status-code-is-supposed-to-be-used-to-tell-the-client-the-session-has/22484262#22484262

However given the frequent usage of Promises with tink,
it might seem a good idea to have this error code for better discrimination.
For instance if one was to sanitize http error messages in Production to
apply those from the error code, having a mere 401 would be problematic.

With a 440 there is not this problem anymore.

emugel commented:
I have a doubt if it shouldn't be in tink.Error as well? 
But let's first see if idea is not rejected.

Commit:
Release 1.3.2

kevinresol commented:
Perhaps we should use an extension type to hold these values. As they may clash each other if we add more (e.g. 499)

Something like this might work:

```haxe
enum abstract MicrosoftStatusCode(Int) {
  var SessionTimeout = 440;
  @:to public function toStatusCode():HttpStatusCode;
  @:to public function toStatusMessage():HttpStatusMessage;
}

```

grepsuzette commented:
Maybe you're right, if we pick non-standard codes here and there and there's a massive standardization of new error codes in a few years it will be painful to handle ; if not this 440 it will be for another one. 

I will see how usable it is with your `MicrosoftStatusCode` idea and come back to comment :)

grepsuzette commented:
It should better go in another file no?
So that we can import as we need

grepsuzette commented:
I have difficulties to switch/case an error from the httpstatus code (it's related because I need this to also test the code I want to add in this PR).

Below is my test file (it is using the current `http-status` lib, not this PR):

```haxe
import httpstatus.HttpStatusCode;                                                                                                                                                               
import tink.core.Error;                                                                                                                                                                         
                                                                                                                                                                                                
class T {                                                                                                                                                                                       
    public static function main() {                                                                                                                                                             
                                                                                                                                                                                                
        // not in ErrorCode                                                                                                     
        var e = new Error(HttpStatusCode.LoopDetected, "");                                                                                                                                     
                                                                                                                                                                                                
        switch e.code {                                                                                                                                                                         
            case Forbidden: 0;                                                                                                                                                                         
            case LoopDetected: 1;   // fails with `httpstatus.HttpStatusCode should be Int`                                                                                                     
            case _: 2;                                                                                                                                                                          
        }                                                                                                                                                                                       
                                                                                                                                                                                                
        switch (e.code : Int) {                                                                                                                                                                 
            case Forbidden: 0;                                                                                                                                                                         
            case LoopDetected.toInt(): 1;  // fails with `Unrecognized pattern: LoopDetected.toInt()` (bug in compiler?)                                                                        
            case (LoopDetected:Int): 1; // fails with `Unrecognized pattern: (LoopDetected:Int)` 
            case _: 2;                                                                                                                                                                          
        }                                                                                                                                                                                       
                                                                                                                                                                                                
                // Now...                                                                                                                                                                       
                // In http-status,                                                                                                                                                              
                // if we modify `abstract HttpStatusCode(Int) from Int` to add ` to Int` implicit conversion:                                                                                   
                                                                                                                                                                                                
                switch e.code {                                                                                                                                                                 
                    case Forbidden: 0;                                                                                                                                                                         
                    case LoopDetected: 1;   // still fails                                                                                                                                      
                    case _: 2;                                                                                                                                                                  
                }                                                                                                                                                                               
                                                                                                                                                                                                
                switch (e.code : Int) {                                                                                                                                                         
                   case Forbidden: 0;                                                                                                                                                                         
                    case LoopDetected: 1;  // works                                                                                                                                             
                    case _: 2;                                                                                                                                                                  
                }                                                                                                                                                                               
                                                                                                                                                                                                
    }                                                                                                                                                                                           
}
```

Think maybe there is a bug in haxe in `case A.toInt():` where `A` is an abstract with a `@:to toInt():Int` function. Otherwise the only way I found is 

1. add implicit conversion `to Int` to `HttpStatusCode` and have a `switch (e.code : Int)`,
2. have a `var nLoopDetected = LoopDetected.toInt()` and then `case nLoopDetected:`, which sucks big time. 

Do you see any other way?

grepsuzette commented:
Maybe can bring @back2dos to the rescue too

grepsuzette commented:
Oh, found one way that works: 
`switch (e.code : HttpStatusCode)`.
Will now see how it can play out with another Microsoft abstract,
and if it works will likely add a small README.md just for the client side

grepsuzette commented:
Something like this compiles:

```haxe
package httpstatus.unofficial;

import httpstatus.HttpStatusCode;
import httpstatus.HttpStatusMessage;

@:enum
abstract Microsoft(HttpStatusCode) from HttpStatusCode {

    var SessionTimeout = 440;

    @:to public function toStatusCode():HttpStatusCode return this;
    @:to public function toStatusMessage():HttpStatusMessage {
        return switch this {
            case 440: 'Client Session Expired Must Log In Again';
            default: 'Unknown Status';
        }   
    }   

    #if tink_core
    @:from public static inline function fromErrorCode(code:tink.core.Error.ErrorCode):Microsoft return HttpStatusCode.fromErrorCode((code:Int));
    #end
}
```

```haxe
class T { 
    public static function main() {
        var e = new Error(HttpStatusCode.LoopDetected, "");
        switch (e.code : HttpStatusCode) {
            case Forbidden: 0;
            case LoopDetected: 1;
            case _: 2;
        }   

        var e = new Error(HttpStatusCode.LoopDetected, "");
        switch (e.code : Microsoft) {
            case SessionTimeout: 0;
            case LoopDetected: 1;
            case _: 2;
        }   
    }   
}
```
Do you see any potential problem?

kevinresol commented:
I am not particularly interested in the conversion between ErrorCode and HttpStatusCode.
ErrorCode is tink specific and _just happen_ that it borrowed some http codes, while http code is standardized.
So, for the sake of this PR. I am more interested in the interop between extension codes and the base status code abstract.

e.g. 
```haxe
import httpstatus.HttpStatusCode;
import httpstatus.Microsoft;
var code:HttpStatusCode = SessionTimeout;
var message:HttpStatusMessage = SessionTimeout;
var message:HttpStatusMessage = code; // this is bit of a stretch...
```

grepsuzette commented:
Ok.
I can get rid of the `@:from` method using this code on the client:
```haxe
var e = new Error(HttpStatusCode.LoopDetected, "");
var errCode : Int = e.code;
switch (errCode : Microsoft) {
    case SessionTimeout: 0;
    case LoopDetected: 1;
    case _: 2;
}
```
In that case no conversion from Error is necessary.
Will push a new commit for review shortly.

Commit:
put SessionTimeout in httpstatus.unofficial.Microsoft

grepsuzette commented:
Oh you mean in this way.
```haxe
import httpstatus.HttpStatusCode;
import httpstatus.Microsoft;
var code:HttpStatusCode = SessionTimeout;
var message:HttpStatusMessage = SessionTimeout;
var message:HttpStatusMessage = code; // this is bit of a stretch...
```
I didn't understand it at first. Well, may see if possible but no guarantee.
The `var code:HttpStatusCode = SessionTimeout;` works already as is, will see if can do anything with the others.

grepsuzette commented:
In fact it already works with case 2 as well:

```haxe
import httpstatus.HttpStatusCode;
import httpstatus.unofficial.Microsoft;

var code:HttpStatusCode = SessionTimeout; trace(code); 
// 440

var message:HttpStatusMessage = SessionTimeout; trace(message); 
// "Client Session Expired Must Log In Again"

var message:HttpStatusMessage = code; trace(code); // this is bit of a stretch...
// "Unknown Status"

var codeMs:Microsoft = SessionTimeout;
var message:HttpStatusMessage = codeMs; trace(message);
// "Client Session Expired Must Log In Again"
```
I have absolutely no idea about how we can avoid the "Unknown Status" when messaging from a code casted to a HttpStatusCode. Lest the 440 is known from HttpStatusMessage but not from HttpStatusCode, in which case the `import` plays no part for that.

Commit:
add @:to Int to httpstatus.unofficial.Microsoft to be able to `new tink.core.Error(SessionTimeout, SessionTimeout)`

kevinresol commented:
> I have absolutely no idea about how we can avoid the "Unknown Status"

I think it is not possible with the current design.

Maybe we can introduce a `HttpStatus` abstract/object to hold both the code and message value together.

grepsuzette commented:
I was proposing that but I edited my message thinking it was crazy. I think if we try it should be in another PR.

grepsuzette commented:
regarding names, is `httpstatus.unofficial.Microsoft` acceptable?

If one day some such codes become standard, we can just take off from whichever `unofficial` class, put them in HttpStatusCode and since underlying type of unofficial is `HttpStatusCode` it should not break code. 

About `Microsoft` I like it for brevity/memorization as I will often use the full class path in that case `httpstatus.unofficial.Microsoft` but we can still change it if that’s not your preferred naming scheme.

Commit:
add missing proprietary Microsoft status codes

Commit:
allow Microsoft unofficial codes to be built from tink core Error codes

supposing we get a tink Error from the network, this allows:

```
switch (e : Microsoft) {
    case SessionTimeout:
}
```

Without this commit, one more line has to be used:
```
var code:Int = e.code;
switch (code : Microsoft) {
    case SessionTimeout:
```

It is also consistent with HttpStatusMessage which has the same explicit conversion.

Commit:
documentation update

Commit:
fix Microsoft.fromErrorCode()

kevinresol commented:
> regarding names, is `httpstatus.unofficial.Microsoft` acceptable?

LGTM

Regarding the tink_core casts. I prefer not to have them here. Maybe you should add `#if http_status` in tink_core repo instead.

grepsuzette commented:
There is the same `@:from` cast in HttpStatusCode, what makes it different for unofficial classes?

kevinresol commented:
Oh well i forgot i put them there but i think i regret it now. Mainly because this library should be purely focus on the http standard without being aware of anything else, and it should be other libraries to reference this lib, not the way around. 

Anyway if it is already there then i think if might be fine to also have it in this PR

grepsuzette commented:
Yes a pure spec library.

For now it is `tink_core` which is ignorant of `httpstatus` (a search command `ack httpstatus` or `ack http_status` gives no result at all). I think it's not that bad the way it is.

kevinresol commented:
Ok let's have this for now. But later I would like to find a proper way to get rid of other library references. Thanks!
